### ISSUE TITLE
Inconsistency between the current and the original implementation of SM-2

### ISSUE THREAD

**Issue by L-M-Sherlock** [eudoxia0/sm2#1]
https://github.com/eudoxia0/sm2/blob/fefd5a50e2efd14f17474491b1b4f80288bb018a/src/lib.rs#L98-L113

It's inconsistent with [for n>2: I(n):=I(n-1)*EF](https://super-memory.com/english/ol/sm2.htm#:~:text=for%20n%3E2%3A%20I(n)%3A%3DI(n%2D1)*EF).

Given this sequence of grades: Good, Good, Hard, Hard, Good

The intervals calculated by the original algorithm are:

1, 6, 6 * 2.5 = 15, 15 * 2.36 = 35.4 -> 35, 35 * 2.22 = 77.7 -> 78

But the current implementation will output:

```
[tests/test.rs:135:9] item.interval() = 1
[tests/test.rs:135:9] item.interval() = 6
[tests/test.rs:135:9] item.interval() = 15
[tests/test.rs:135:9] item.interval() = 30
[tests/test.rs:135:9] item.interval() = 66
```

Here is my code for test:

```rust
#[test]
fn test_max_interval() {
    let mut item = Item::default();
    let qs = [
        Quality::Good,
        Quality::Good,
        Quality::Hard,
        Quality::Hard,
        Quality::Good,
    ];
    for q in qs {
        item = item.review(q);
        dbg!(item.interval());
    }
}
```

Btw, the `let i = i.ceil()` should be replaced with `let i = i.round()` because [If interval is a fraction, round it up to the nearest integer.
](https://super-memory.com/english/ol/sm2.htm#:~:text=If%20interval%20is%20a%20fraction%2C%20round%20it%20up%20to%20the%20nearest%20integer.)

**Comment by eudoxia0** [comment]
Thanks!

This is actually a double bug: I didn't realize the algorithm description required rounding, but also, I used `ceil` when I meant to use `floor`, on the intuition that, whether an interval is `10.1` or `10.9` days, it should be floored down to "ten days".

**Comment by eudoxia0** [comment]
> The intervals calculated by the original algorithm are:
>
>1, 6, 6 * 2.5 = 15, 15 * 2.36 = 35.4 -> 35, 35 * 2.22 = 77.7 -> 78

Ah, I see. I realize there's a conceptual ambiguity. There's two possible ways to read the description of the algorithm:

1. `I(n-1)` is the recorded value of the last interval.
2. `I(n-1)` is a recursive call to the interval function, using the current value of `EF`.

So my closed-form expansion assumes the EF is constant, where, of course, the whole point is that EF changes. That's interesting.

**Comment by L-M-Sherlock** [comment]
> Ah, I see. I realize there's a conceptual ambiguity.

Yep. The original description is ambiguous. The source code is more clear: [Interval:=round(Interval*EF);
](https://super-memory.com/english/ol/sm2source.htm#:~:text=Interval%3A%3Dround(Interval*EF)%3B), where the `Interval` is read from `TDataRecord`.

**Comment by eudoxia0** [comment]
Fascinating. Thank you. I'm gonna think about how to rewrite this cleanly.
